pkg/controller/workloadspread/workloadspread_controller.go

```diff
@@ -0,0 +1,714 @@
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(controllerName, mgr, controller.Options{
+		Reconciler: r, MaxConcurrentReconciles: concurrentReconciles,
+		RateLimiter: ratelimiter.DefaultControllerRateLimiter()})
+	if err != nil {
+		return err
+	}
+
+	// Watch for spec changes to WorkloadSpread
+	err = c.Watch(&source.Kind{Type: &appsv1alpha1.WorkloadSpread{}}, &handler.EnqueueRequestForObject{},
+		predicate.GenerationChangedPredicate{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Pods have a specific annotation
+	err = c.Watch(&source.Kind{Type: &corev1.Pod{}}, &podEventHandler{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for replica changes to CloneSet
+	err = c.Watch(&source.Kind{Type: &appsv1alpha1.CloneSet{}}, &workloadEventHandler{Reader: mgr.GetCache()})
+	if err != nil {
+		return err
+	}
+
+	// Watch for replica changes to Deployment
+	err = c.Watch(&source.Kind{Type: &appsv1.Deployment{}}, &workloadEventHandler{Reader: mgr.GetCache()})
+	if err != nil {
+		return err
+	}
+
+	// Watch for replica changes to ReplicaSet
+	err = c.Watch(&source.Kind{Type: &appsv1.ReplicaSet{}}, &workloadEventHandler{Reader: mgr.GetCache()})
+	if err != nil {
+		return err
+	}
+
+	// Watch for parallelism changes to Job
+	err = c.Watch(&source.Kind{Type: &batchv1.Job{}}, &workloadEventHandler{Reader: mgr.GetCache()})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	cli := util.NewClientFromManager(mgr, controllerName)
+	return &ReconcileWorkloadSpread{
+		Client:   cli,
+		scheme:   mgr.GetScheme(),
+		recorder: mgr.GetEventRecorderFor(controllerName),
+	}
+}
+
+var _ reconcile.Reconciler = &ReconcileWorkloadSpread{}
+
+// ReconcileWorkloadSpread reconciles a WorkloadSpread object
+type ReconcileWorkloadSpread struct {
+	client.Client
+	scheme   *runtime.Scheme
+	recorder record.EventRecorder
+}
+
+// +kubebuilder:rbac:groups=apps.kruise.io,resources=workloadspreads,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups=apps.kruise.io,resources=workloadspreads/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=apps.kruise.io,resources=clonesets,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps,resources=replicasets,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;create;update;patch;delete
```

Comment:
Please remove the useless verbs for deployments/replicasets/jobs/pods, such as create/delete/patch/update?